### PR TITLE
Add support for halo updates in TileCommunicator, use in dycore

### PR DESCRIPTION
**Pull request by mcgibbon**
## Purpose

This PR adds support for halo updates to TileCommunicator, and modifies the dycore so it can accept either a TileCommunicator or a CubedSphereCommunicator.

This enables doubly-periodic domains.

## Code changes:

- Added the following attributes/methods to Communicator base class: `tile`, `halo_update`, `boundaries`, `start_halo_update`, `vector_halo_update`, `start_vector_halo_update`, `synchronize_vector_interfaces`, `start_synchronize_vector_interfaces`, `get_scalar_halo_updater`, and `get_vector_halo_updater`
- Communicator is now an abc.ABC abstract base class
- Added `boundary` and `tile_index` methods to the Partitioner abstract base class
- Deleted deprecated `finish_halo_update` method from CubedSphereCommunicator

## Checklist
Before submitting this PR, please make sure:

- [x] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
- [x] Docstrings and type hints are added to new and updated routines, as appropriate
- [x] All relevant documentation has been updated or added (e.g. README, CONTRIBUTING docs)
- [x] For each public change and fix in `pace-util`, HISTORY has been updated
- [x] Unit tests are added or updated for non-stencil code changes

### CODE REVIEW

**Comment by mcgibbon** [comment]
Jenkins failure appears to be a random failure due to the distributed filesystem issue with gt4py.

File: pace-util/pace/util/communicator.py [hunk 1]
```diff
@@ -645,3 +518,211 @@ def get_vector_halo_updater(
...
+
+        Args:
+            quantity: the quantity to be updated
+            n_points: how many halo points to update, starting from the interior
+
+        Returns:
+            request: an asynchronous request object with a .wait() method
+        """
+        if self.partitioner.layout[0] < 3 or self.partitioner.layout[1] < 3:
+            raise NotImplementedError()
```
Comment (FlorianDeconinck) [inline pace-util/pace/util/communicator.py:578]: Maybe a comment here to explain why it's not implemented

**Review by FlorianDeconinck — APPROVED** [review]
LGTM

File: pace-util/pace/util/communicator.py (hunk 1 above)
Comment (FlorianDeconinck) [reply pace-util/pace/util/communicator.py:578]: (and below)

Comment (mcgibbon) [reply pace-util/pace/util/communicator.py]: Thanks, updated!

**Pull request merged**